### PR TITLE
TTNN Sharded tensor support

### PR DESCRIPTION
**Pull request by ntarafdar**
This adds sharded tensor support at the ttnn level.

This also simplifies the API by integrating shard_spec into memory_config. This unifies the API such that we don't need to explicitly include memory_config. 

This also pulls in the changes in this PR: https://github.com/tenstorrent-metal/tt-metal/pull/4520
(refactor shard spec underneath hood to have the buffer level include the tensor level shard spec)

### CODE REVIEW

**Comment by tarafdarTT** [comment]
@TT-billteng @DrJessop , this cherry-picks the other PR you guys reviewed

File: tt_metal/impl/buffers/buffer.hpp
Comment (TT-billteng) [inline tt_metal/impl/buffers/buffer.hpp:89]: can you mark these member functions as `const`?

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (TT-billteng) [inline tt_metal/impl/dispatch/command_queue.cpp:358]: would it make sense to create a new class `EnqueueReadShardedBufferCommand` for shard specific code ?

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py
Comment (mywoodstock) [inline tests/tt_eager/python_api_testing/unit_testing/test_sharded.py:535]: why remove bfp8_b testing?

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py
Comment (mywoodstock) [inline tests/tt_eager/python_api_testing/unit_testing/test_sharded.py:531]: why remove testing of `out_sharded = true`?

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py
Comment (mywoodstock) [inline tests/tt_eager/python_api_testing/unit_testing/test_sharded.py:567]: why remove bias?

File: tt_eager/tt_dnn/op_library/operation.hpp
Comment (mywoodstock) [inline tt_eager/tt_dnn/op_library/operation.hpp:221]: what is this mysterious number? can you add a comment about it?

File: ttnn/tensor.py
Comment (mywoodstock) [inline ttnn/tensor.py:322]: Can the mem config be changing the shard type? like input is height sharded, and want to change to block sharded, etc.?

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (DrJessop) [inline tt_metal/impl/dispatch/command_queue.cpp:330]: not a fan of the tabbing + no space between ( and {

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (DrJessop) [reply tt_metal/impl/dispatch/command_queue.cpp:358]: I agree. I think EnqueueReadBuffer/EnqueueWriteBuffer can behind the scenes call the relevant API.

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (DrJessop) [inline tt_metal/impl/dispatch/command_queue.cpp:430]: as @TT-billteng suggested, I think we should make new commands for sharding, and have enqueue_read_buffer create the correct command.

File: tt_metal/impl/dispatch/command_queue.hpp
Comment (DrJessop) [inline tt_metal/impl/dispatch/command_queue.hpp:100]: If you do what Bill suggested, we won't need an assemble_device_command_sharded. It'll just be the assemble_device_command for EnqueueWrite/ReadBuffer

**Review by DrJessop — CHANGES_REQUESTED** [review]
I think as @TT-billteng suggested, we should create EnqueueRead/WriteShardedBufferCommand classes, and have enqueue_read_buffer/enqueue_write_buffer figure out which of those commands to construct and process. This will get rid of the need for pre-amble/post-amble and new sharded device command APIs.

File: tests/ttnn/unit_tests/test_matmul.py
Comment (arakhmati) [inline tests/ttnn/unit_tests/test_matmul.py:316]: Does `[M // grid_size[0], K // grid_size[1]]` needed as an extra argument? Can we ever pass in any other values?

**Review by arakhmati — CHANGES_REQUESTED** [review]

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py
Comment (tarafdarTT) [reply tests/tt_eager/python_api_testing/unit_testing/test_sharded.py]: My mistake, didn't mean to push that.

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py
Comment (tarafdarTT) [reply tests/tt_eager/python_api_testing/unit_testing/test_sharded.py]: My mistake, didn't mean to push that.

File: tests/tt_eager/python_api_testing/unit_testing/test_sharded.py
Comment (tarafdarTT) [reply tests/tt_eager/python_api_testing/unit_testing/test_sharded.py]: My mistake, didn't mean to push that.

File: tt_eager/tt_dnn/op_library/operation.hpp
Comment (tarafdarTT) [reply tt_eager/tt_dnn/op_library/operation.hpp:221]: I think it's equivalent to the size in bytes of Memory_Config. I am not sure however, @arakhmati is that the case?

File: ttnn/tensor.py
Comment (tarafdarTT) [reply ttnn/tensor.py:322]: I can add that capability, if we are supplying a memconfig of a sharded tensor, and the tensor is already sharded.
Underneath the hood we can do sharded_to_interleaved and interleaved_to_sharded

File: ttnn/tensor.py
Comment (arakhmati) [inline ttnn/tensor.py:108]: Why is decorator `decorate_operation()` used here?

File: tt_eager/tt_dnn/op_library/operation.hpp
Comment (arakhmati) [reply tt_eager/tt_dnn/op_library/operation.hpp:221]: DeviceOperation uses type erasure to store the underlying operation struct. And the storage must be as big as the biggest underlying operation struct.

**Review by eyonland — CHANGES_REQUESTED** [review]
We need to include documentation to the matmul itself about the sharding configuration once we figure out how we want the api to work here. 

Also, is one test giving us adequate coverage here?

**Comment by tarafdarTT** [comment]
> We need to include documentation to the matmul itself about the sharding configuration once we figure out how we want the api to work here.
> 
> Also, is one test giving us adequate coverage here?

I added another test to cover another sharding method.
TBH I am not totally familiar with the matmul implementation underneath the hood (I reverse engineered the parameters to feed in by looking at the asserts). @TT-BrianLiu could you look at adding some documentation regarding sharding for the matmul?

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (DrJessop) [inline tt_metal/impl/dispatch/command_queue.cpp:371]: See my command about naming this function in device_command.cpp

**Review by DrJessop — APPROVED** [review]
Approving so as not to block you, but I believe addressing my new comments should be quite quick and would improve the code readability.

**Review by arakhmati — APPROVED** [review]

File: tests/ttnn/unit_tests/test_matmul.py
Comment (mywoodstock) [inline tests/ttnn/unit_tests/test_matmul.py:330]: shouldn't this be `(12, 8)`? We use `x, y` for grids

**Comment by TT-BrianLiu** [comment]
Why do we need `ttnn.create_height_sharded_memory_config`, `ttnn.create_width_sharded_memory_config`, and `ttnn.create_block_sharded_memory_config`? All we are hiding is just one `ShardingStrategy` variable. I think we just need `create_sharded_memory_config`. 

Also, I don't think we should have a default for shard orientation. User should be conscious of this since matmul has requirements on shard orientation. (Unless we are okay with errors being thrown back at the user).

**Comment by TT-BrianLiu** [comment]
Can we also directly do:
```
input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout, sharded_memory_config)
```
instead of:
```
input_tensor_a = ttnn.from_torch(torch_input_tensor_a)
input_tensor_a = ttnn.to_device(input_tensor_a, device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
input_tensor_a = ttnn.to_layout(input_tensor_a, ttnn.TILE_LAYOUT)
input_tensor_a = ttnn.to_memory_config(input_tensor_a, sharded_mem_config)
```

**Comment by tarafdarTT** [comment]
> Can we also directly do:
> 
> ```
> input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout, sharded_memory_config)
> ```
> 
> instead of:
> 
> ```
> input_tensor_a = ttnn.from_torch(torch_input_tensor_a)
> input_tensor_a = ttnn.to_device(input_tensor_a, device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
> input_tensor_a = ttnn.to_layout(input_tensor_a, ttnn.TILE_LAYOUT)
> input_tensor_a = ttnn.to_memory_config(input_tensor_a, sharded_mem_config)
> ```

Currently we also do support

```
input_tensor_a = ttnn.from_torch(torch_input_tensor_a)
input_tensor_a = ttnn.to_device(input_tensor_a, device, sharded_mem_config)
```

Underneath the hood though it does something slightly different.
The example you posted (going to device and then to_memory_config) copies to device as an interleaved buffer (`to_device`), and then `to_memory_config` does the `interleaved_to_sharded`.

The example I posted below, moves directly to device as a sharded buffer. Our infra supports that if we use the sharded_mem_config instead. I wrote the matmul examples using interleaved_to_sharded to be consistent with examples we already have in tt_sharded.

I think we should leave both paths as users may want to either move to interleaved first and then convert or straight to sharded. 

Whether or not we want to combine `to_device` and `from_torch` in one command, that I think is a separate discussion. Maybe outside the scope of this PR.

**Comment by tarafdarTT** [comment]
hey @eyonland just waiting for your approval. the scope of this PR, is to add sharding support at the TTNN level, I have added a couple matmul examples but I don't think its super thorough. @TT-BrianLiu is working on a subsequent PR working on this PR that will have more support for matmul.

**Review by boris-drazic — APPROVED** [review]

File: tests/ttnn/unit_tests/test_matmul.py
Comment (tarafdarTT) [reply tests/ttnn/unit_tests/test_matmul.py:356]: @mywoodstock at the TTNN level y,x are swapped when inputting core_grid. I decided to leave my sharding grid to be consistent with the rest of ttnn

**Comment by arakhmati** [comment]
> Can we also directly do:
> 
> ```
> input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout, sharded_memory_config)
> ```
> 
> instead of:
> 
> ```
> input_tensor_a = ttnn.from_torch(torch_input_tensor_a)
> input_tensor_a = ttnn.to_device(input_tensor_a, device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
> input_tensor_a = ttnn.to_layout(input_tensor_a, ttnn.TILE_LAYOUT)
> input_tensor_a = ttnn.to_memory_config(input_tensor_a, sharded_mem_config)
> ```



@TT-BrianLiu 

Yes, we can do that now:
```
input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout=layout, device=device, memory_config=sharded_memory_config)
```

But as @tarafdarTT

**Comment by davorchap** [comment]
> > Can we also directly do:
> > ```
> > input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout, sharded_memory_config)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > instead of:
> > ```
> > input_tensor_a = ttnn.from_torch(torch_input_tensor_a)
> > input_tensor_a = ttnn.to_device(input_tensor_a, device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
> > input_tensor_a = ttnn.to_layout(input_tensor_a, ttnn.TILE_LAYOUT)
> > input_tensor_a = ttnn.to_memory_config(input_tensor_a, sharded_mem_config)
> > ```
> 
> @TT-BrianLiu
> 
> Yes, we can do that now:
> 
> ```
> input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout=layout, device=device, memory_config=sharded_memory_config)
> ```
> 
> But as @tarafdarTT mentioned, we want to keep different paths

having both paths of doing is great

**Review by davorchap — APPROVED** [review]

**Review by farbabi — APPROVED** [review]

File: docs/source/ttnn/tensor.rst
Comment (eyonland) [inline docs/source/ttnn/tensor.rst:151]: Replace 
```
``Layout``
```
with
```
:ref:`Layout<ttnn.Layout>`
```

File: docs/source/ttnn/tensor.rst
Comment (eyonland) [inline docs/source/ttnn/tensor.rst:155]: Can we create a link to page (like we have for ttnn.Layout) that explains what a page is?

**Review by eyonland — CHANGES_REQUESTED** [review]
Great work!    Love this documentation.   Just two minor changes and it looks good to go!

File: docs/source/ttnn/tensor.rst
Comment (tarafdarTT) [reply docs/source/ttnn/tensor.rst]: hey eon, sorry I meant a tile, which we already explain above.

File: docs/source/ttnn/tensor.rst
Comment (tarafdarTT) [reply docs/source/ttnn/tensor.rst]: fixed that here: https://github.com/tenstorrent-metal/tt-metal/pull/4597/commits/1c8231d97e2faba7aa786ce9f7630fbd2fd325a0

**Review by arakhmati — APPROVED** [review]

**Review by eyonland — APPROVED** [review]

File: docs/source/ttnn/tensor.rst
Comment (mywoodstock) [inline docs/source/ttnn/tensor.rst:161]: The above seems to imply that sharding only works with `TILE`. I think sharding description should be independent of the layout, and then mention that both TILE and RM layouts can be used.

File: docs/source/ttnn/tensor.rst
Comment (tarafdarTT) [reply docs/source/ttnn/tensor.rst:151]: the user doesn't do either, this is for us to have a conceptual understanding of sharding so the user can understand the diagrams that come after.

File: docs/source/ttnn/tensor.rst
Comment (tarafdarTT) [reply docs/source/ttnn/tensor.rst:161]: this is to remain consistent with the rest of the documentation where in the layout section we already defined a tile. I do not want to confuse the user more by having them differentiate between tiles and pages.

File: tt_metal/impl/dispatch/command_queue.hpp
Comment (TT-billteng) [inline tt_metal/impl/dispatch/command_queue.hpp:90]: why is this needed?

File: tt_metal/impl/dispatch/command_queue.hpp
Comment (TT-billteng) [inline tt_metal/impl/dispatch/command_queue.hpp:171]: why is this needed?

**Review by farbabi — APPROVED** [review]

**Review by TT-billteng — APPROVED** [review]

**Pull request merged**
